### PR TITLE
#6374: Add a form to configure min and max scale denominator range visibility inside the layer settings panel

### PR DESCRIPTION
**Pull request by allyoucanmap**
## Description

This PR adds a form to layer settings to update min and max resolutions values implemented in the PR https://github.com/geosolutions-it/MapStore2/pull/6407

Preview
image

The form includes:

- toggle button to enable/disable visibility limits
- max value select input
- min value select input
- limits type select input
- synch button to get min and max scale denominator from WMS GetCapabilities (available only on layer with wms type)

This PR replaces also the Dropdown (react-widget) components in the Display tab with react-select

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

## Issue

**What is the current behavior?**

#6374

**What is the new behavior?**

This PR add an UI to update min and max resolutions values of a layer object

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [x] No

## Other useful information

merge after https://github.com/geosolutions-it/MapStore2/pull/6407

[map 32620](https://dev.mapstore2.geo-solutions.it/mapstore/#/viewer/openlayers/32620) for test

### CODE REVIEW

File: web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+                onChange(isNaN(newValue) ? undefined : { ...option, value: newValue });
+            }}
+        />
+    );
+}
+/**
+ * Form to set min and max resolutions values from layer object
+ * @memberof components.TOC
+ * @name VisibilityLimitsForm
+ * @class
```
Comment (MV88) [inline web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx:85]: it's not a class, so this can be removed, right?

File: web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx [hunk 2]
```diff
@@ -0,0 +1,405 @@
...
+                    if (isMounted.current) {
+                        const {
+                            maxScaleDenominator,
+                            minScaleDenominator
+                        } = capabilities;
+                        const newMaxValue = !isNil(maxScaleDenominator)
+                            && getResolutionObject(maxScaleDenominator, 'scale');
+                        const newMinValue = !isNil(minScaleDenominator)
+                            && getResolutionObject(minScaleDenominator, 'scale');
+                        const newMessageId = isNil(maxScaleDenominator) && isNil(minScaleDenominator)
```
Comment (MV88) [inline web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx:185]: the message is quite misleading imo

https://user-images.githubusercontent.com/11991428/105369424-d3eac300-5c02-11eb-9dcd-f344fd027d61.mp4

why the fields remains empty if it says "has been updated correctly"?

File: web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx [hunk 3]
```diff
@@ -0,0 +1,405 @@
...
+    const isMounted = useRef();
+
+    useEffect(() => {
+        isMounted.current = true;
+        return () => {
+            isMounted.current = false;
+        };
+    }, []);
+
+    function clearAlters() {
```
Comment (MV88) [inline web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx:164]: my 2 cents on this name
```suggestion
    function clearMessages() {
```
or
```suggestion
    function clearAlerts() {
```

File: web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx [hunk 4]
```diff
@@ -0,0 +1,405 @@
...
+                .catch(() => {
+                    if (isMounted.current) {
+                        setLoading(false);
+                        setError('layerProperties.visibilityLimitsValuesError');
+                    }
+                });
+        }
+    }
+
+    useEffect(() => {
```
Comment (MV88) [inline web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx:205]: this effect is executed when, limitsType or projection changes, but they are not used inside the effect.
maybe i'm wrong here, but can you clarify?

File: web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx [hunk 5]
```diff
@@ -0,0 +1,405 @@
...
+                        value: option[limitsType] + (option.delta || 0),
+                        label: option[limitsType],
+                        showArrow: option.zoom === zoom,
+                        disabled: !isNil(minValue?.[limitsType]) && option[limitsType] <= minValue[limitsType],
+                        limitsType
+                    };
+                })}
+                disabled={disableResolutionLimits || loading}
+                onChange={(option) => {
+                    // created entries do not contain zoom propery
```
Comment (MV88) [inline web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx:291]:
```suggestion
                    // created entries do not contain zoom property
```

File: web/client/plugins/tocitemssettings/defaultSettingsTabs.js [hunk 6]
```diff
@@ -42,6 +43,9 @@ const responses = {
 import { StyleSelector } from '../styleeditor/index';
 
 const StyleList = defaultProps({ readOnly: true })(StyleSelector);
+const ConnectedDisplay = connect(
+    createSelector([mapSelector], ({ zoom, projection }) => ({ zoom, projection }))
```
Comment (MV88) [inline web/client/plugins/tocitemssettings/defaultSettingsTabs.js:47]: can this be a problem if toc is used outside our store? 
i mean what do you think to pass zoom and projection from the TocItemSettingsPlugin.jsx [here](https://github.com/geosolutions-it/MapStore2/blob/master/web/client/plugins/TOCItemsSettings.jsx#L108)?
if not, just ignore this comment

File: web/client/translations/data.en-US.json [hunk 7]
```diff
@@ -104,6 +104,21 @@
             "templateFormatInfoAlert2": "Use ${ attribute } to wrap the properties you need to display",
             "templateFormatInfoAlertExample": "The id of the feature is ${ properties }",
             "templatePreview": "Template preview",
+            "visibilityLimitsTitle": "Visibility limits",
```
Comment (MV88) [inline web/client/translations/data.en-US.json:107]: probably these could have been an object 
"visibilityLimits": {
            "title": "Visibility limits",

but is fine as you did too, feels free to keep them as is

**Review by MV88 — CHANGES_REQUESTED** [review]
Some small fixes required below, we can discuss them no big deal.

i have although find out that GFI requests are not ignoring layers outside visibility limits.
image


Print is printing the layer even if it is outside visibilityLimits

File: web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx (hunk 4 above)
Comment (allyoucanmap) [reply web/client/components/TOC/fragments/settings/VisibilityLimitsForm.jsx:205]: After a double check those dependencies were not correct because the right ones should be dpu and resolutions. Going to fix this

File: web/client/translations/data.en-US.json (hunk 7 above)
Comment (allyoucanmap) [reply web/client/translations/data.en-US.json:107]: I'm keeping current structure

File: web/client/translations/data.it-IT.json [hunk 8]
```diff
@@ -104,6 +104,21 @@
...
             "templateFormatInfoAlertExample": "id della feature è ${ properties }",
             "templatePreview": "Anteprima template",
+            "visibilityLimitsTitle": "Limiti di visibilità",
+            "visibilityLimitsValuesUpdateUndefined": "I valori di scala minimo e massimo sono stati aggiornati correttamente ma sono risultati entrambi non definiti",
+            "visibilityLimitsValuesUpdate": "I valori di scala minimo e massimo sono stati aggiornati correttamente",
+            "visibilityLimitsValuesError": "Non è possibile richiedere i valori di scala minimo e massimo via richiesta WMS GetCapabilities",
+            "visibilityLimitsUpdateWMSCapabilities": "Ottini i valori di scala minimi e massimi con una richiesta WMS GetCapabilities. Questa richiesta sincronizzerà anche i valori di scala minimo e massimo non definiti",
+            "visibilityLimitsMaxValue": "Valore massimo (escluso)",
+            "visibilityLimitsMaxValuePlaceholder": "Seleziona il valore massimo",
+            "visibilityLimitsMaxValueNoResultsText": "Nessuna opzione corrisponde al valore inserito",
```
[the remark this answers is not in the document]
Comment (allyoucanmap) [reply web/client/translations/data.it-IT.json:114]: I'll merge in a sinlge message visibilityLimitsValueNoResultsText

File: web/client/utils/MapUtils.js [hunk 9]
```diff
@@ -205,6 +206,15 @@ export function getScales(projection, dpi) {
     return getResolutions().map((resolution) => resolution * dpu);
 }
 
+export function getZoomFromResolution(targetResolution, resolutions = getResolutions()) {
```
[the remark this answers is not in the document]
Comment (allyoucanmap) [reply web/client/utils/MapUtils.js:209]: Those are comments related to the specific lines of code.
I'll had an additional jsdocs for the function:

```
/**
 * Convert a resolution to the nearest zoom
 * @param {number} targetResolution resolution to be converted in zoom
 * @param {array} resolutions list of all available resolutions
 */
```

File: web/client/translations/data.en-US.json (hunk 7 above)
Comment (allyoucanmap) [reply web/client/translations/data.en-US.json:107]: @MV88 I'll follow your suggestion, please ignore my previous comment

**Review by MV88 — APPROVED** [review]

**Comment by tdipisa** [comment]
@ElenaGallo please test this in DEV and let us know when we can do the backport.

**Comment by allyoucanmap** [comment]
> The synch button to get min and max scale denominator from WMS GetCapabilities does not works (see the gs:ny_poi on this map)

@tdipisa @ElenaGallo 

- Issue 1
The synch button gets the min and max scales informations from the Layer object of the capabilities

image

This layer has 3 styles associated and some of those does not have min max scale denominator. The WMS capabilities seems to compute min and max scales from all the styles available to cover all the possible configuration of the layer + style (this is my understanding but maybe it's better to ask to some geoserver dev).
I think for this reason there are no min max values in the capabilities.

- Issue 2

Probably in this case there were two min max resolutions limits declared on the client that does not match with the style min and max scale denominator. 


in conclusion I think the synch capabilities works well when a layer has a single style or all styles have the same scale denominator limits

**Comment by tdipisa** [comment]
@ElenaGallo the above points you reported are related to a problem with the selected style "Point of interest" (poi.sld). Probably the style has been edited a second time in GeoServer and the style's metadata have not been updated too. 

image

Is MS that menages style's metadata through the style editor, therefore it is not a problem of this min/max scale tool but rather a style editor issue. Did you modify the style directly in GeoServer? If yes can you please edit the style entirelly MS side for your tests.
Let us know.

@allyoucanmap can you anyway open an issue for an improvement of the current style editor? MS loads both the metadata and the real style from GeoServer, therefore it should compare them and give the user the choice of which one to use with a message (something like "There are modifications performed on the style outside MS. Do you want to overwrite them or use the version previously saved with MapStore?)

**Comment by ElenaGallo** [comment]
@tdipisa  I had not  changed the style from geoserver but now I tried to change the style from MS and I noticed it wasn't updated.
Anyway by modifying the style from MS through the style editor the min/max scale tool works correctly

**Comment by tdipisa** [comment]
@ElenaGallo can we proceed with the backport then?

**Comment by ElenaGallo** [comment]
@tdipisa yes, @allyoucanmap please backport this PR on a stanle branch
